src/extra/AutoUI/Lenses/index.tsx

```diff
@@ -12,13 +12,17 @@ import { IconProp } from '@fortawesome/fontawesome-svg-core';
 import map from 'lodash/map';
 import uniq from 'lodash/uniq';
 
-export interface LensTemplate {
+export interface LensTemplate<T = any> {
 	slug: string;
 	name: string;
 	data: {
 		label: string;
 		format: string;
-		renderer: (props: any) => React.ReactElement;
+		renderer: (
+			props:
+				| types.CollectionLensRendererProps<T>
+				| types.EntityLensRendererProps<T>,
+		) => React.ReactElement;
```

Comment:
I wouldn't pass *either* CollectionLensRendererProps *or* EntityLensRendererProps. I would pass a list of all possible props that might be passed (see the following for all possible props that might be passed: https://github.com/balena-io-modules/rendition/blob/17cc165fa308a714d4a832d57b5f555c6f92ded6/src/extra/AutoUI/index.tsx#L389-L399)
The reason I would not restrict it to *either* Collection *or* Entity is because the user is able to create their own CustomLenses and we do not want to force them to use any props they do not need, so most or all of these props that we pass into `lens.data.renderer` are likely optional and the user should be free to use them all.

Reply:
This is not an exclusive OR in the typings. It just dictates that the AutoUI component is always guaranteed to be passing those props, depending on how the component is used. That's is a bit less vague for a consumer trying to create a lens. With this they know they what props we are guaranteeing to be setting depending on how the lens is used.

If we made all those props optional, then this will add extra unnecessary checks in the Lens' implementation, checking for each prop whether its defined or not, which with this you can be sure that as far your `fitler` `type` is respected, then you will always be passed the above props.
And it's not that the Lens will have to support all those props, it's just that it will have to use the props interface like:
```
export const table2: LensTemplate = {
	slug: 'table',
	name: 'Default table lens',
	data: {
		label: 'Table',
		format: 'table',
		renderer: (props: CollectionLensRendererProps<number>) => (
			<>{props.data?.length ?? null}</>
		),
		icon: faTable,
		type: '*',
		filter: {
			type: 'array',
			items: {
				type: 'number',
			},
		},
	},
};
```
Moreover for mixed components, you can reason the move in which the Lense is used by just checking one differentiating property and then for the rest of the implementation TS will only give you the correct types & props available in that mode.